File: document/src/vespa/document/select/result.cpp
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "result.h"
#include <ostream>

namespace document::select {

Result Result::Invalid;
Result Result::False;
Result Result::True;

Result::Result() = default;

const Result&
Result::operator!() const {
    if (this == &Invalid) {
        return Invalid;
    }
    return (this == &True ? False : True);
}

const Result&
Result::operator&&(const Result& r) const {
    if (this == &False || &r == &False) {
        return False;
    }
    return (this == &True && &r == &True ? True : Invalid);
}

const Result&
Result::operator||(const Result& r) const {
    if (this == &True || &r == &True) {
        return True;
    }
    if (this == &Invalid || &r == &Invalid) {
        return Invalid;
    }
    return False;
}

void
Result::print(std::ostream& out, bool,
              const std::string&) const
{
    if (this == &Invalid) out << "Invalid";
    else if (this == &True) out << "True";
    else out << "False";
}

}

File: document/src/main/java/com/yahoo/document/select/Result.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.document.select;

import com.yahoo.document.select.rule.AttributeNode;

/**
 * @author Simon Thoresen Hult
 */
public enum Result {

    /**
     * Defines all enumeration constants.
     */
    TRUE,
    FALSE,
    INVALID;

    @Override
    public String toString() {
        return name().toLowerCase();
    }
    
    /**
     * Inverts the result value to the appropriate value. True → False, False → True and Invalid → Invalid.
     * @return inverted result
     */
    public static Result invert(Result result) {
        if (result == Result.TRUE) return Result.FALSE;
        if (result == Result.FALSE) return Result.TRUE;
        return Result.INVALID;
    }

    /**
     * Converts the given object value into an instance of this Result enumeration.
     *
     * @param value The object to convert.
     * @return The corresponding result value.
     */
    public static Result toResult(Object value) {
        if (value == null || value == Result.FALSE || value == Boolean.FALSE ||
            (value instanceof Number && ((Number)value).doubleValue() == 0)) {
            return Result.FALSE;
        } else if (value == INVALID) {
            return Result.INVALID;
        } else if (value instanceof AttributeNode.VariableValueList) {
            return ((AttributeNode.VariableValueList)value).isEmpty() ? Result.FALSE : Result.TRUE;
        } else if (value instanceof ResultList) {
            return ((ResultList)value).toResult();
        } else {
            return Result.TRUE;
        }
    }
}
